### PR TITLE
oop-special_class_method-hard-q12

### PR DESCRIPTION
**Pull request by KaiKhademi**
## PR Checklist
- [x]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [x]  Are the assertion fail messages good?
    - [x]  Logical (does it make sense?)
    - [x]  Readable (can a student understand what’s wrong?)
    - [x]  Not too feed-y (does it give away the answer?)
    - [x]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [x]  Specific (does the fail message relate to the question?)
- [x]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [x]  EOF newlines
- [x]  Do the comments for Parsons follow the correct convention?
- [x]  Is the question being tested sufficiently?
    - [x]  Are there representative cases to test the program?
    - [x]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [x]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”

### CODE REVIEW

File: src/oop/special_class_method/hard/q12/QuestionText.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# oop-special_class_method-hard-q12
+
+## Question Text
+
+You are developing a program that teaches people about the physical traits of a variety of human-made objects. Write a 
+Java class called **Laptop** that has two attributes, an `int` called **yearMade** and a `String` called **texture**
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q12/QuestionText.md:6]: texture does not seem related to laptops

File: src/oop/special_class_method/hard/q12/MainTest.java [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+    private ObjectTest testClass;
+
+    @BeforeEach
+    public void setUp() throws Throwable {
+        String classString = "oop.special_class_method.hard.q12." + className;
+        this.testClass = new ObjectTest(classString);
+    }
+
+    @Test
+    public void alienClassHasCorrectAttributes() {
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q12/MainTest.java:34]: change this and all other test titles to Laptop instead of Alien

File: src/oop/special_class_method/hard/q12/MainTest.java [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+        assertTrue(testClass.hasField(attributeName1),
+                "Your " + className + " class is missing the " + attributeName1 + " attributes, it is spelled incorrectly.");
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q12/MainTest.java:36]:
```suggestion
                "Your " + className + " class is missing the " + attributeName1 + " attributes, or it is spelled incorrectly.");
```

File: src/oop/special_class_method/hard/q12/MainTest.java [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+                "Your " + getAttributeMethodName1 + " method does not return the value of the " + attributeName1 + " attribute.");
+        getMethodOutput = testClass.callMethod(getAttributeMethodName2, testInstance);
+        _assertEquals("Initial String", getMethodOutput,
+                "Your " + getAttributeMethodName2 + " method does not return the value of the " + attributeName2 + " attribute.");
+        Object[][] setMethodArguments = {
+                {value1, int.class}
+        };
+        testClass.callMethod(setAttributeMethodName1, setMethodArguments, testInstance);
+        _assertEquals(value1, testClass.callMethod(getAttributeMethodName1, testInstance),
+                "Your " + getAttributeMethodName1 + " method does not correctly update the value of " + setAttributeMethodName1 + " after using the " + getAttributeMethodName1 + " method.");
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q12/MainTest.java:205]: Fix this and the below error message. the first getAttributeMethodName and setAttributeMethodName should be replaced by setAttributeMethodName and then attributeName

**Review by SamamthaH — CHANGES_REQUESTED** [review]

File: src/oop/special_class_method/hard/q12/MainTest.java [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+    @BeforeEach
+    public void setUp() throws Throwable {
+        String classString = "oop.special_class_method.hard.q12." + className;
+        this.testClass = new ObjectTest(classString);
+    }
+
+    @Test
+    public void laptopClassHasCorrectAttributes() {
+        assertTrue(testClass.hasField(attributeName1),
+                "Your " + className + " class is missing the " + attributeName1 + " attributes, or it is spelled incorrectly.");
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q12/MainTest.java:36]: change "attributes" to singular for both variable messages

File: src/oop/special_class_method/hard/q12/MainTest.java [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+
+    @ParameterizedTest
+    @MethodSource("inputProvider")
+    public void correctGetYearMadeMethod(int value1, String value2) throws Throwable {
+        Object[][] arguments = {
+                {value1, int.class},
+                {value2, String.class}
+        };
+        Object testInstance = testClass.createInstance(arguments);
+        Object getMethodOutput = testClass.callMethod(getAttributeMethodName1, testInstance);
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q12/MainTest.java:114]: use .hasMethod first here and everywhere else

File: src/oop/special_class_method/hard/q12/MainTest.java [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+        };
+        Object testInstance = testClass.createInstance(instantiationArguments);
+        Object[][] setMethodArguments = {
+                {updatedValue, int.class}
+        };
+        Class<?>[] methodModifierClasses = {
+                int.class
+        };
+        assertTrue(testClass.hasMethod(setAttributeMethodName1, methodModifierClasses, Void.TYPE),
+                "Your " + setAttributeMethodName1 + " method does not have the correct name, return type, or parameters.");
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q12/MainTest.java:151]: split this assert into two so the return value can have it's own assert. Do this in the other set method test as well.

File: src/oop/special_class_method/hard/q12/QuestionText.md [hunk 8]
```diff
@@ -0,0 +1,27 @@
+# oop-special_class_method-hard-q12
+
+## Question Text
+
+You are developing a program that teaches people about the physical traits of a variety of human-made objects. Write a 
+Java class called **Laptop** that has two attributes, an `int` called **yearMade** and a `String` called **texture** 
+that are initialized by the constructor
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q12/QuestionText.md:7]: get rid of duplicate class information

File: src/oop/special_class_method/hard/q12/QuestionText.md [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+Your `Laptop` class must have the following items:
+
+- Attributes:
+    - An `int` called **yearMade**
+    - A `String` called **colour**
+    - These are not visible from other classes
+
+- Constructor:
+    - a `Laptop` constructor that is visible to other classes and accepts an `int` and a `String` and uses it to initialize
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q12/QuestionText.md:19]:
```suggestion
    - a `Laptop` constructor that is visible to other classes and accepts an `int` and a `String` and uses them to initialize
```

**Review by SamamthaH — CHANGES_REQUESTED** [review]

File: src/oop/special_class_method/hard/q12/QuestionText.md [hunk 10]
```diff
@@ -0,0 +1,26 @@
+# oop-special_class_method-hard-q12
+
+## Question Text
+
+You are developing a program that teaches people about the physical traits of a variety of human-made objects. Write a 
+Java class called **Laptop**.
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q12/QuestionText.md:6]:
```suggestion
Java class called **Laptop** to model an object using the requirements below:
```

File: src/oop/special_class_method/hard/q12/QuestionText.md [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+
+- Constructor:
+    - a `Laptop` constructor that is visible to other classes and accepts an `int` and a `String` and uses them to initialize
+      the **yearMade** and **colour** attributes.
+
+- Methods:
+    - a `setYearMade` method which accepts an `int` and uses it to set the value of the **yearMade** attribute
+    - a `getYearMade` method which returns the value stored in the **yearMade** attribute
+    - a `setColour` method which accepts a `String` and uses it to set the value of the **colour** attribute
+    - a `getColour` method which returns the value stored in the **colour** attribute
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q12/QuestionText.md:25]: Specify the return type for set methods.

File: src/oop/special_class_method/hard/q12/MainTest.java [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+
+
+    @ParameterizedTest
+    @MethodSource("inputProvider")
+    public void correctGetYearMadeMethod(int value1, String value2) throws Throwable {
+        Object[][] arguments = {
+                {value1, int.class},
+                {value2, String.class}
+        };
+        Object testInstance = testClass.createInstance(arguments);
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q12/MainTest.java:123]: Move this call below the method check.

File: src/oop/special_class_method/hard/q12/MainTest.java [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+    }
+
+    @ParameterizedTest
+    @MethodSource("inputProvider")
+    public void correctGetColourMethod(int value1, String value2) throws Throwable {
+        Object[][] arguments = {
+                {value1, int.class},
+                {value2, String.class}
+        };
+        Object testInstance = testClass.createInstance(arguments);
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q12/MainTest.java:152]: Same thing here.

File: src/oop/special_class_method/hard/q12/MainTest.java [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+    }
+
+    @ParameterizedTest
+    @MethodSource("updateIntegerInputProvider")
+    public void correctSetYearMadeMethod(int initialiValue, String value2, int updatedValue) throws Throwable {
+        Object[][] instantiationArguments = {
+                {initialiValue, int.class},
+                {value2, String.class}
+        };
+        Object testInstance = testClass.createInstance(instantiationArguments);
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q12/MainTest.java:181]: And here

File: src/oop/special_class_method/hard/q12/MainTest.java [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+        testClass.callMethod(setAttributeMethodName1, setMethodArguments, testInstance);
+        _assertEquals(value1, testClass.callMethod(getAttributeMethodName1, testInstance),
+                "Your " + getAttributeMethodName1 + " method does not correctly update the value of " + attributeName1 + " after using the " + setAttributeMethodName1 + " method.");
+        setMethodArguments = new Object[][]{
+                {value2, String.class}
+        };
+        testClass.callMethod(setAttributeMethodName2, setMethodArguments, testInstance);
+        _assertEquals(value2, testClass.callMethod(getAttributeMethodName2, testInstance),
+                "Your " + getAttributeMethodName2 + " method does not correctly update the value of " + attributeName2 + " after using the " + setAttributeMethodName2 + " method.");
+    }
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q12/MainTest.java:258]: I would call the set method before the get method here, not the other way around.

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/oop/special_class_method/hard/q12/MainTest.java [hunk 16]
```diff
@@ -0,0 +1,257 @@
...
+
+    @Test
+    public void getYearMadeIsDefinedCorrectly() {
+        String incorrectMethodDefinition = "The %s method in the %s class is not defined correctly. Make sure it is declared, spelt correctly, and has the correct parameters.";
+        String incorrectModifierMessage = "The %s method in the %s class has the wrong visibility modifier.";
+        String incorrectReturnType = "The %s method in the %s class has the incorrect return type.";
+        assertTrue(testClass.hasMethod(getAttributeMethodName1, null), String.format(incorrectMethodDefinition, getAttributeMethodName1, className));
+        assertTrue(testClass.hasModifier(getAttributeMethodName1, null, "public"), String.format(incorrectModifierMessage, getAttributeMethodName1, className));
+        assertTrue(testClass.hasReturnType(getAttributeMethodName1, null, int.class), String.format(incorrectReturnType, getAttributeMethodName1, className));
+    }
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q12/MainTest.java:113]: You can remove the tests for "x method is defined correctly"

File: src/oop/special_class_method/hard/q12/MainTest.java [hunk 17]
```diff
@@ -0,0 +1,257 @@
...
+                Arguments.of(1989, "Black"),
+                Arguments.of(2020, "Blue"),
+                Arguments.of(2017, "White"),
+                Arguments.of(2010, "Yellow")
+        );
+    }
+
+    private static Stream<Arguments> updateIntegerInputProvider() {
+        return Stream.of(
+                Arguments.of(2020, "Red", 2021),
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q12/MainTest.java:62]: Should add test for setting the value to 0

File: src/oop/special_class_method/hard/q12/MainTest.java [hunk 18]
```diff
@@ -0,0 +1,257 @@
...
+                Arguments.of(2019, "Blue", 2001),
+                Arguments.of(2013, "White", 2007),
+                Arguments.of(1999, "Yellow", 2006)
+
+        );
+    }
+
+    private static Stream<Arguments> updateStringInputProvider() {
+        return Stream.of(
+                Arguments.of(3, "Red", "White"),
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q12/MainTest.java:73]: Should add case to set string to ""

File: src/oop/special_class_method/hard/q12/MainTest.java [hunk 19]
```diff
@@ -0,0 +1,257 @@
...
+        assertTrue(testClass.hasField(attributeName2, String.class),
+                "Your " + attributeName2 + " does not have the correct type.");
+        assertTrue(testClass.hasModifier(attributeName2, "private"),
+                "Your " + attributeName2 + " does not have the correct visibility modifier.");
+    }
+
+
+    private static Stream<Arguments> inputProvider() {
+        return Stream.of(
+                Arguments.of(2002, "Red"),
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q12/MainTest.java:52]: Should add 0 and "" as test case as you use this for the methods work together test.

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]
